mesa/experimental/devs/eventlist.py

```diff
@@ -0,0 +1,191 @@
+
+    TODO:: can also be used for agents.unique_id
+    """
+
+    def __init__(cls, name, bases, attrs):
+        super().__init__(name, bases, attrs)
+        cls._ids = itertools.count(1)
+
+
+class Priority(IntEnum):
+    LOW = 10
+    DEFAULT = 5
+    HIGH = 1
+
+
+class SimulationEvent(metaclass=InstanceCounterMeta):
+    """A simulation event
+
+    the callable is wrapped using weakrefs, so there is no need to explicitly cancel event if e.g., an agent
+    is removed from the simulation.
+
+    Attributes:
+        time (float): The simulation time of the event
+        function  (Callable): The function to execute for this event
+        priority (Priority): The priority of the event
+        unique_id (int) the unique identifier of the event
+        function_args (list[Any]): Argument for the function
+        function_kwargs (Dict[str, Any]): Keyword arguments for the function
+
+    """
+
+    @property
+    def CANCELED(self) -> bool:
+        return self._canceled
+
+    def __init__(
+        self,
+        time: int | float,
+        function: Callable,
+        priority: Priority = Priority.DEFAULT,
+        function_args: list[Any] | None = None,
+        function_kwargs: dict[str, Any] | None = None,
+    ) -> None:
+        super().__init__()
+        self.time = time
+        self.priority = priority.value
+        self._canceled = False
+
+        if isinstance(function, MethodType):
+            function = WeakMethod(function)
+        else:
+            function = ref(function)
+
+        self.fn = function
+        self.unique_id = next(self._ids)
+        self.function_args = function_args if function_args else []
+        self.function_kwargs = function_kwargs if function_kwargs else {}
+
+        if self.fn is None:
+            raise Exception()
+
+    def execute(self):
+        """execute this event"""
+        fn = self.fn()
+        if fn is not None:
+            fn(*self.function_args, **self.function_kwargs)
+
+    def cancel(self) -> None:
+        """cancel this event"""
+        self._canceled = True
+        self.fn = None
+        self.function_args = []
+        self.function_kwargs = {}
+
+    def __cmp__(self, other):
+        if self.time < other.time:
+            return -1
+        if self.time > other.time:
+            return 1
+
+        if self.priority > other.priority:
+            return -1
+        if self.priority < other.priority:
+            return 1
+
+        if self.unique_id < other.unique_id:
+            return -1
+        if self.unique_id > other.unique_id:
+            return 1
+
+        # theoretically this should be impossible unless it is the
+        # exact same event
+        return 0
+
+    def _to_tuple(self):
+        return self.time, self.priority, self.unique_id, self
+
+
+class EventList:
+    """An event list
+
+    This is a heap queue sorted list of events. Events are allways removed from the left. The events are sorted
```

Comment:
happy to add the info.

In case of pydsol the main reason for this extra info is that dsol (java) was using a different data structure (this has changed now as far as I know).

Reply:
An info stating the diff between the DEVS design choice in this PR and pydsol would be helpful. It seems that your implementation of EventList is simpler here, or at least, the ordering scheme is clearer.